lib/msgpack/timestamp.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module MessagePack
+  Timestamp = Struct.new(:sec, :nsec)
+
+  class Timestamp # a.k.a. "TimeSpec"
+    TYPE = -1 # timestamp extension type
+
+    TIMESTAMP32_MAX_SEC = 0x100000000 # 32-bit unsigned int
+    TIMESTAMP64_MAX_SEC = 0x400000000 # 34-bit unsigned int
+
+    def self.from_msgpack_ext(data)
+      case data.length
+      when 4
+        # timestamp32 = (uint32be)
+        sec = data.unpack('L>').first
+        new(sec, 0)
+      when 8
+        # timestamp64 = (uint32be, uint32be)
+        n, s = data.unpack('L>2')
+        sec = (n & 0x3) * 0x100000000 + s
```

Comment:
Can you use `0b11` as the mask and `<< 32` instead of `* 0x100000000`?
It's easy to read for me because my recommendation shows "lower 2bit mask" and "use it as upper bytes". Counting the number of zero in `0x100000000` is scary.

Reply:
And... this is the just an idea, which might be a kind of too-early optimization.
How about this code? `sec = (n == 0 ? s : (n & 0b11 << 32) + s)`
Until 2038, `n` is 0 always (or in most cases).

Reply:
I think its optimization is a bit more  complicated like this:

```ruby
sec_hi_2bits = n & 0b11
sec = sec_hi_2bits == 0 ? s : (sec_hi_2bits << 32) + s
```

I'm not sure it is faster than just `(n & 0b11) << 32) + s`.

Reply:
Ah, your code looks right. And here's my microbenchmark:
```ruby
require 'benchmark'

t = Time.now
n = t.nsec << 2
s = t.sec

times = 1_000_000
Benchmark.bm do |x|
  x.report(:simple) do
    for i in 1..times
      sec = (n & 0b11 << 32) + s
    end
  end
  x.report(:optimized) do
    for i in 1..times
      sec_hi_2bits = n & 0b11
      sec = sec_hi_2bits == 0 ? s : (sec_hi_2bits << 32) + 2
    end
  end
end
```
The result was:
```
       user     system      total        real
simple  0.055794   0.000129   0.055923 (  0.056197)
optimized  0.047181   0.000082   0.047263 (  0.047393)
```
Let's forget this idea...